File: apps/web/utils/orgPositionAssignee.ts
import type { EmployeeInfo, OrgPosition } from '../types';

/** Активные сотрудники на должности (по карточкам HR); иначе legacy — holderUserId. */
export function getMemberUserIdsForPosition(
  position: OrgPosition | undefined,
  employees: EmployeeInfo[]
): string[] {
  if (!position) return [];
  const fromCards = employees
    .filter((e) => !e.isArchived && e.orgPositionId === position.id && e.userId)
    .map((e) => e.userId);
  const uniq = [...new Set(fromCards)].sort();
  if (uniq.length > 0) return uniq;
  if (position.holderUserId) return [position.holderUserId];
  return [];
}

export type PositionAssigneeResolution = {
  assigneeId: string | null;
  assigneeIds?: string[];
  /** Обновить должность (курсор round-robin) */
  positionPatch?: Partial<OrgPosition>;
};

/**
 * Назначение задачи на шаг с assigneeType === 'position':
 * - round_robin — одна задача, исполнитель меняется по кругу (по умолчанию);
 * - all — одна задача с assigneeIds = все участники (ответственный в UI — первый).
 */
export function resolveAssigneesForOrgPosition(
  position: OrgPosition | undefined,
  employees: EmployeeInfo[]
): PositionAssigneeResolution {
  if (!position) return { assigneeId: null };

  const members = getMemberUserIdsForPosition(position, employees);
  if (members.length === 0) return { assigneeId: null };

  const mode = position.taskAssigneeMode === 'all' ? 'all' : 'round_robin';

  if (mode === 'all') {
    return {
      assigneeId: members[0],
      assigneeIds: members,
    };
  }

  if (members.length === 1) {
    return { assigneeId: members[0] };
  }

  const last = position.lastTaskAssigneeUserId;
  const idx = last ? members.indexOf(last) : -1;
  const next = members[(idx + 1) % members.length];
  return {
    assigneeId: next,
    positionPatch: { lastTaskAssigneeUserId: next },
  };
}

File: apps/bot/taska_bot/domain/org_position_assignee.py
"""Согласовано с apps/web/utils/orgPositionAssignee.ts — назначение задач на должность."""
from __future__ import annotations

from typing import Any, Optional, TypedDict


class PositionAssigneeResolution(TypedDict, total=False):
    assignee_id: Optional[str]
    assignee_ids: list[str]
    position_patch: dict[str, Any]


def get_member_user_ids_for_position(
    position: Optional[dict], employees: list[dict]
) -> list[str]:
    if not position:
        return []
    pid = str(position.get("id") or "")
    from_cards: list[str] = []
    for e in employees:
        if e.get("isArchived"):
            continue
        if str(e.get("orgPositionId") or "") == pid and e.get("userId"):
            from_cards.append(str(e["userId"]))
    uniq = sorted(set(from_cards))
    if uniq:
        return uniq
    holder = position.get("holderUserId")
    if holder:
        return [str(holder)]
    return []


def resolve_assignees_for_org_position(
    position: Optional[dict], employees: list[dict]
) -> PositionAssigneeResolution:
    if not position:
        return {"assignee_id": None}

    members = get_member_user_ids_for_position(position, employees)
    if not members:
        return {"assignee_id": None}

    mode = position.get("taskAssigneeMode") or "round_robin"
    if mode == "all":
        return {"assignee_id": members[0], "assignee_ids": members}

    if len(members) == 1:
        return {"assignee_id": members[0]}

    last = position.get("lastTaskAssigneeUserId")
    idx = members.index(last) if last and last in members else -1
    nxt = members[(idx + 1) % len(members)]
    return {
        "assignee_id": nxt,
        "position_patch": {"lastTaskAssigneeUserId": nxt},
    }
